src/rgw/rgw_qos.h

```diff
@@ -0,0 +1,246 @@
+#pragma once
+#include <chrono>
+#include <thread>
+#include <condition_variable>
+#include "rgw_common.h"
+
+#if !defined(dout_subsys)
+#define dout_subsys ceph_subsys_rgw
+#define def_dout_subsys
+#endif
+
+#define dout_context g_ceph_context
```

Comment:
it's best not to mess with dout macros in header files. i don't think QosDatastruct needs to do any logging. for QosActiveDatastruct, you can either move its implementation into a rgw_qos.cc file and use the macros there, or add a `DoutPrefixProvider*` argument to its constructor and do the logging with `ldpp_dout()` which doesn't depend on the `dout_subsys`/`dout_context` macros

Reply:
I think we would want to get logging from QosDatastruct(Ratelimiter), maybe with even higher logging level (30) so in case of bug in the rate limiting the user will be able to bring us logs that could help us debug it.
I will add to both classes the DoutPrefixProvider*

Reply:
ok. i think `QosActiveDatastruct` should be its own provider since it's running a background thread, and all of the log output should look the same. to do that,
```c++
class QosActiveDatastruct : public DoutPrefix {
 public:
  QosActiveDatastruct(CephContext* cct)
    : DoutPrefix(cct, ceph_subsys_rgw, "rate limiter: ") {
```
for logging inside of `QosDatastruct`, we'll be calling this from the context of an `RGWOp`, so want to use that op's prefix. that means the caller will need to pass in the `const DoutPrefixProvider* dpp` as an argument to increase_entry()/increase_bw() (and not to its constructor)

log statements within `QosActiveDatastruct` would use `ldpp_dout(this, 20) <<`, and `ldpp_dout(dpp, 20) <<` in `QosDatastruct`